### PR TITLE
Nginx change - use more modern cipher

### PR DESCRIPTION
**Pull request by henokgetachew**
# Description

#8335

The list of ciphers in the nginx config are outdated. This PR updates them.

# Code review checklist
- [x] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:nginx-cipher-fix/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:nginx-cipher-fix/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:nginx-cipher-fix/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: nginx/nginx.conf
Comment (mrjones-plip) [inline nginx/nginx.conf:46]: Mozilla ("intermediate", nginx `1.19.6`, openssl `1.1.1g`) [suggest](https://ssl-config.mozilla.org/#server=nginx&version=1.19.6&config=intermediate&openssl=1.1.1g&guideline=5.7) `off` as does rando [SF post](https://serverfault.com/questions/997614/setting-ssl-prefer-server-ciphers-directive-in-nginx-config):

```suggestion
        ssl_prefer_server_ciphers   off;
```

File: nginx/nginx.conf
Comment (mrjones-plip) [inline nginx/nginx.conf:49]: yes! so great to see `v1` and `v1.1` get dropped!  Welcome `v1.3`, lookin' good ; )

File: nginx/nginx.conf
Comment (mrjones-plip) [inline nginx/nginx.conf:50]: I'm not super up on my ciphers, but is this a random list or specific?  I'd maybe [defer to Mozilla](https://ssl-config.mozilla.org/#server=nginx&version=1.19.6&config=intermediate&openssl=1.1.1g&guideline=5.7), but would want to be careful to not orphan any old-ish, but still `TLSv1.2` compatible browsers.

**Review by mrjones-plip — CHANGES_REQUESTED** [review]
Yay! Super great to see some security forward motion. Thank you!

Looks like only TLS 1.3 is a LOOOONG way off (API Level 29+ [needed](https://developer.android.com/reference/javax/net/ssl/SSLSocket.html), which is only [on Android 10+](https://apilevels.com/)). 1.2 is a great improvement though!

What about HTTP2 though? I think this would be a nice addition and is a nice backwards compatible change we can add at the top of the `server {` stanza (IPv6 too, but...meh):

```
server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;
```

Finally, HSTS would be a good addition too:

```
    add_header Strict-Transport-Security "max-age=63072000" always;
```

I'm trying to be careful to not break backwards compatibility, but lemme know if I'm missing anything!

**Comment by mrjones-plip** [comment]
Some notes on my review of this PR:

* I tested with this config:

**Comment by mrjones-plip** [comment]
Sorry! clicked "Close with Comment" and very much did not mean to!!

**Comment by mrjones-plip** [comment]
aha! I see the API is [adding the HSTS header upstream](https://github.com/medic/cht-core/blob/2d9702c131a4ff054b4769e839aea47e7fadb190/api/package-lock.json#L4901), that's why there's two and why we don't need to add it in nginx.

**Comment by henokgetachew** [comment]
Thanks @dianabarsan! Yes let's include it in the release.

**Comment by dianabarsan** [comment]
Thanks @henokgetachew 
Do you have the bandwidth to address the feedback from @mrjones-plip ?

**Comment by henokgetachew** [comment]
I think I can address the comments on Friday of this week. Is that too late?

**Comment by dianabarsan** [comment]
Friday is great! Thanks a lot!

File: nginx/nginx.conf
Comment (henokgetachew) [reply nginx/nginx.conf]: Okay I am meeting you half way here. I have added the older ones we used to support to the current list.

File: nginx/nginx.conf
Comment (henokgetachew) [reply nginx/nginx.conf:46]: I can't say I agree here. This has always been `on` in this config. Only reason to turn it `off` seems to be to support older clients. Given we've added the ciphers back, I think it makes sense to leave it `on`.

**Comment by mrjones-plip** [comment]
Hey all! I was out of holiday for month of July so not sure if this still needs any input from me. Please ping me if more review is needed!

**Comment by garethbowen** [comment]
NB: I've force pushed a rebase to get the build to pass. Sorry about that!

**Comment by garethbowen** [comment]
@mrjones-plip I would appreciate it if you could review this again, given that you've been so active so far, to make sure you're happy with where this ended up. I note that the http2 change didn't make it - I propose we raise a new issue for this to unblock this PR.

**Comment by mrjones-plip** [comment]
@garethbowen  - will do! I'll get to this by week's end.

File: nginx/nginx.conf
Comment (mrjones-plip) [reply nginx/nginx.conf]: @henokgetachew - what is the logic for meeting us half way instead of going with the more [explicit Mozilla list](https://ssl-config.mozilla.org/#server=nginx&version=1.19.6&config=intermediate&openssl=1.1.1g&guideline=5.7)? While I don't feel super strongly, the whole point of this ticket is to update the ciphers so I figure we may as well dig into this a bit!

my advice is to follow Mozilla with:

`ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305;`

While we're in there, do `ssl_prefer_server_ciphers off;` as well.

Otherwise, if you have explicit reasons why this is a bad idea, please lemme know!

**Comment by mrjones-plip** [comment]
I re-tested this on Android 5.1 with CHT Android 1.0.x, Firefox latest, Chrome Latest,  and Chrome 113.  Just wanting to firm up our cipher selection [per above](https://github.com/medic/cht-core/pull/8278#discussion_r1283672932) before approving the PR.

File: nginx/nginx.conf
Comment (garethbowen) [reply nginx/nginx.conf]: I agree with @mrjones-plip here. I'm by no means an expert but the advice seems to be if you're dropping TLS 1.1 and below, then you should turn `ssl_prefer_server_ciphers` off because it lets the client pick their preferred cypher which can improve performance on the device. Because these two settings are related it makes sense to change the `ssl_prefer_server_ciphers` off at the same time.

**Comment by garethbowen** [comment]
@mrjones-plip I've aligned this with the mozilla recommendations. Would you mind reviewing again? Do you feel that you've done sufficient testing?

@nydr I've also included your `msie6` change - thanks!

**Review by mrjones-plip — APPROVED** [review]
LGTM!

I tested by pulling in the latest nginx image (first had to track down the right tag! lookin' at you `public.ecr.aws/s5s3h4s7/cht-nginx:4.3.0-nginx-cipher-fix`) and spun up a CHT Core 4.2.2 instance.  I pointed my Android 5.1/WebView 95.0 device with CHT Android 1.0.4 at the local-ip URL which uses a Let's Encrypt cert.  I verified I could log in with an offline user and poke around a bit.

Here's a screenshot of the "about" page from the device:

image

The only thing I could think would be to have @medic/quality-assurance  test a few more phones or possibly have @latin-panda have a look given all her WebView work for CHT 4.0.0 she did.   Since I tested on WebView `95.0` ,  testing on older version , but that are still supported would be ideal!

However, I suspect we're fine as is!

**Comment by mrjones-plip** [comment]
follow up link to [http/2 ticket](https://github.com/medic/cht-core/issues/8451)

**Comment by mrjones-plip** [comment]
Adding another validation point that with `nmap` the current `nginx` looks like this (note TLS 1.0 and 1.1 and short cipher list):

```
/snap/bin/nmap --script +ssl-enum-ciphers 192-168-68-26.local-ip.medicmobile.org -p 10444
Starting Nmap 7.93 ( https://nmap.org ) at 2023-08-11 08:41 PDT
Nmap scan report for 192-168-68-26.local-ip.medicmobile.org (192.168.68.26)
Host is up (0.00016s latency).
rDNS record for 192.168.68.26: 192-168-68-26.my.local-ip.co

PORT      STATE SERVICE
10444/tcp open  unknown
| ssl-enum-ciphers: 
|   TLSv1.0: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (ecdh_x25519) - A
|     compressors: 
|       NULL
|     cipher preference: indeterminate
|     cipher preference error: Too few ciphers supported
|   TLSv1.1: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (ecdh_x25519) - A
|     compressors: 
|       NULL
|     cipher preference: indeterminate
|     cipher preference error: Too few ciphers supported
|   TLSv1.2: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (ecdh_x25519) - A
|     compressors: 
|       NULL
|     cipher preference: indeterminate
|     cipher preference error: Too few ciphers supported
|_  least strength: A

Nmap done: 1 IP address (1 host up) scanned in 0.49 seconds
```

and the `nginx` in this branch looks like this (note only TLS 1.2 and 1.3 and longer cipher list):

```
/snap/bin/nmap --script +ssl-enum-ciphers 192-168-68-26.local-ip.medicmobile.org -p 10452        
Starting Nmap 7.93 ( https://nmap.org ) at 2023-08-11 08:34 PDT
Nmap scan report for 192-168-68-26.local-ip.medicmobile.org (192.168.68.26)
Host is up (0.00015s latency).
rDNS record for 192.168.68.26: 192-168-68-26.my.local-ip.co

PORT      STATE SERVICE
10452/tcp open  unknown
| ssl-enum-ciphers: 
|   TLSv1.2: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (secp384r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (secp384r1) - A
|       TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (secp384r1) - A
|     compressors: 
|       NULL
|     cipher preference: client
|   TLSv1.3: 
|     ciphers: 
|       TLS_AKE_WITH_AES_128_GCM_SHA256 (secp384r1) - A
|       TLS_AKE_WITH_AES_256_GCM_SHA384 (secp384r1) - A
|       TLS_AKE_WITH_CHACHA20_POLY1305_SHA256 (secp384r1) - A
|     cipher preference: client
|_  least strength: A

Nmap done: 1 IP address (1 host up) scanned in 0.45 seconds
```

**Comment by tatilepizs** [comment]
I tried to test it and it looks like it is stuck, I am just seeing a spinner before accessing the login page.

Here are some details about the test:

I am using these files:  
- [cht-core.yml](https://github.com/medic/cht-core/files/12323851/cht-core.yml.zip)
- [couchdb.yml](https://github.com/medic/cht-core/files/12323852/couchdb.yml.zip)

Android Version: 9
WebView version: 74.0.3729.136
CHT Android version: 1.0.1

Here are some screenshots about what I am seeing:

<details>
<summary>Console tab</summary>

image

</details>

<details>
<summary>Security tab</summary>

image


</details>

<details>
<summary>Network tab</summary>

image

</details>

If there is any more information that it will be helpfull please let me know

**Comment by lorerod** [comment]
I have the same outcome as @tatilepizs. I'm seeing the spinner and not able to access the login page.
I confirm that my local setup was working by connecting successfully with another phone with WebView 111.

**Environment:** local using yml files in the PR
**Platform:** Mobile App. APK cht-android-v1.0.4-1-unbranded-armeabi-v7a-release
**Phone:** Samsung Galaxy J7
**Android version:** 8.1
**Android System WebView Version:** 69.0.3497.100

<details>
<summary>Console tab</summary>

Captura de pantalla 2023-08-11 a la(s) 13 40 35

</details>

<details>
<summary>Network tab</summary>

Captura de pantalla 2023-08-11 a la(s) 13 40 17

</details>

<details>
<summary>Security tab</summary>

Captura de pantalla 2023-08-11 a la(s) 13 40 25

</details>


<details>
<summary>nginx logs</summary>

```
172.22.0.1 - - [11/Aug/2023:17:35:57 +0000] "GET / HTTP/1.1" 200 414 "-" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
172.22.0.1 - - [11/Aug/2023:17:35:57 +0000] "GET /runtime.js HTTP/1.1" 200 768 "https://192-168-86-49.my.local-ip.co/" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
172.22.0.1 - - [11/Aug/2023:17:35:57 +0000] "GET /polyfills.js HTTP/1.1" 200 19042 "https://192-168-86-49.my.local-ip.co/" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
172.22.0.1 - - [11/Aug/2023:17:35:57 +0000] "GET /scripts.js HTTP/1.1" 200 50741 "https://192-168-86-49.my.local-ip.co/" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
172.22.0.1 - - [11/Aug/2023:17:35:57 +0000] "GET /styles.css HTTP/1.1" 200 77763 "https://192-168-86-49.my.local-ip.co/" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
172.22.0.1 - - [11/Aug/2023:17:35:57 +0000] "GET /main.js HTTP/1.1" 200 1017028 "https://192-168-86-49.my.local-ip.co/" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
172.22.0.1 - - [11/Aug/2023:17:35:57 +0000] "GET /fonts/NotoSans-Regular.ttf HTTP/1.1" 200 221787 "https://192-168-86-49.my.local-ip.co/styles.css" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
172.22.0.1 - - [11/Aug/2023:17:35:59 +0000] "GET /favicon.ico HTTP/1.1" 200 1511 "https://192-168-86-49.my.local-ip.co/" "Mozilla/5.0 (Linux; Android 8.1.0; SM-J710MN Build/M1AJQ; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 org.medicmobile.webapp.mobile/v1.0.4-1"
```

</details>

**Comment by mrjones-plip** [comment]
Thanks so much @tatilepizs  and @lorerod !  

Thinking on this some more, it'd be great if we could iteratively walk back the nginx changes to see which one actually broke the client working.  Looking [here](https://github.com/medic/cht-core/pull/8278/files), we'd try just making one change like `ssl_prefer_server_ciphers  off;`  -> revert to ` ssl_prefer_server_ciphers  on;` , restart nginx and try again etc.

Here's the steps to do it.  Before getting started, check out `nginx-cipher-fix` , so you have `cht-core/nginx/nginx.conf` file local.

1. Start CHT Core instance, test on your mobile device to ensure the error is there
2. In your local editor of choice revert ONE item from the change.  (e.g.  `ssl_prefer_server_ciphers  off;`  -> revert to ` ssl_prefer_server_ciphers  on;`)
3. `cd` into the `cht-core` repo and copy the updated config file into the nginx container.  Here's the command for me after figuring container is `8278_cipher_nginx_1` so copy would be : `docker cp nginx/nginx.conf 8278_cipher_nginx_1:/etc/nginx/nginx.conf`
4. restart nginx: `docker restart 8278_cipher_nginx_1`
5. test again

If we do this one line at a time, we'll figure out exactly which change is breaking the older android.  

We could also try it the other way, copy in [the file from `master`](https://github.com/medic/cht-core/blob/master/nginx/nginx.conf), and then add in each change.  This way step one would be "ensure it works first" ;)

**Comment by mrjones-plip** [comment]
Latest: Android 9/WebView 74/CHT Android 1.0.1 failed the same way as this branch ([spinner w/ JS error](https://github.com/medic/cht-core/pull/8278#issuecomment-1675154934)) when testing against CHT Core 4.2.2.  This is very likely an `nginx` container regression introduced early in 4.x, but wasn't caught until now. 

We're going to go back and test against earlier CHT Core 4.x releases to see when the bug was introduced so we can narrow down what changes to focus on as breaking.

**Comment by garethbowen** [comment]
@mrjones-plip We're confident now that this is a webapp issue and nothing to do with this change right? Is there any more testing required on this change?

**Comment by mrjones-plip** [comment]
@garethbowen  - well...we _[just_ raised the WebView](https://github.com/medic/cht-core/issues/8456#issuecomment-1681321031) to make these older phones obsolete for <4.2, so...yeah, I think this one is good to go as it's already tested well on newer WebViews!

**Pull request merged**
